[6.x] Driver based image manipulation with render tag

jasonvarga opened the pull request:
> [!WARNING]
> This is very much a work in progress.

Currently, all image manipulation is fairly tightly integrated with Glide.

This PR introduces the concept of driver-based image manipulation. Glide will continue to ship as the default, but it will become easier to swap it out for other services such as Imgix, Cloudinary, Cloudflare Images, etc.

This will also introduce a new `render` tag which will be essentially identical to the `glide` tag but allow you to use any driver. For example, if you were to use the imgix driver, you'd be able to use their face detection and background removal features:

```antlers
{{ render:image
    w="100" h="200"
    fit="facearea" faceindex="1" facepad="2"
    bg-remove="true" }}
```

This is aimed to be introduced in the 5.x line at some point, so everything should be backwards compatible. You will be able to opt into it by either using the new `render` tag, or by customizing your image manipulation config file.

```php
return [
    'default' => env('STATAMIC_IMAGE_MANIPULATION_DRIVER', 'glide'),

    'manipulators' => [
        'glide' => [
            'driver' => 'glide',
            'library' => 'gd',
            'cache' => public_path('img'),
        ],

        'imgix' => [
            'driver' => 'imgix',
            'url' => 'myaccount.imgix.net',
            'key' => env('IMGIX_KEY'),
        ],
    ],
];
```

- [ ] Imgix with multiple asset container handling
- [ ] Preset "unpacking". Presets are a Glide feature, but they make sense anywhere. We could continue to let you define presets but instead of passing `?p=whatever` as a param, it would unpack them and do `?w=100&h=100&etc=`
- [ ] Parameter aliases (e.g. `width` => `w`)
- [x] The magical `fit="crop_focal"` param only works for glide. Needs to be smarter. e.g. glide wants `?fit=crop-1-1-1` but imgix wants something like `?fit=crop&crop=focalpoint&fp-x=1&fp-y=1&fp-z=1` 
- [ ] Container specific drivers,
	- [ ] Log/error appropriately depending on environment
- [ ] Resolve https://github.com/statamic/cms/issues/8199
- [ ] Include https://github.com/statamic/cms/pull/10594 somehow

jasonvarga committed:
render tag basics

jasonvarga committed:
dont assume url will be part of attributes

jasonvarga committed:
glide manipulator

jasonvarga committed:
stub imgix

jasonvarga committed:
cp thumbnails use new manipulator and redirects

jasonvarga committed:
existing image generator just offloads work onto new one (which was basically copy pasted)

jasonvarga committed:
deprecations

jasonvarga committed:
Merge branch '5.x' into image-manipulation

jasonvarga committed:
Merge branch '5.x' into image-manipulation

jasonvarga committed:
point to right class

jasonvarga committed:
attributes

jasonvarga committed:
handle tag pair with asset

jasonvarga committed:
wip

jasonvarga committed:
merge in asset data

jasonvarga committed:
manipulator manager - ability to define them in config and extend with new drivers

jasonvarga committed:
render tag can specify a manipulator via the "using" param

jasonvarga committed:
use an enum

jasonvarga committed:
pointing to wrong classes

jasonvarga committed:
wrong method

jasonvarga committed:
initial imgix driver

jasonvarga committed:
imgix dependency

jasonvarga committed:
url builder returns 1s which doesnt work. hack it in for now.

jasonvarga committed:
Merge branch '5.x' into image-manipulation

jasonvarga committed:
phpunit attributes

jasonvarga committed:
Source classes ...

Dedicated classes to represent sources now replace
- awkward string|Asset union types
- SourceType enum. the class itself is the identifier of the type.

All source classes can return an asset.
- For external urls, they will be null
- For assets obviously the asset
- For asset ids it will resolve the asset by id
- For paths it will attempt to resolve by url from the root. This happens for the existing implementation.

Render tag:
- The driver method accepts the raw source and converts it to the class.
- Uses the new source ahead of time to resolve the focal point when using fit="crop_focal"

jasonvarga committed:
Merge branch '5.x' into image-manipulation

jasonvarga committed:
fix source here

jasonvarga committed:
null

godismyjudge95 commented:
Would this be something in the future where Statamic itself would support a certain number of popular image providers by default?  Or would they all be addon based?  Just curious because we use Cloudflare images as our image provider.

jasonvarga commented:
Cloudflare images is probably a good one to also support natively.

jasonvarga committed:
Merge branch '5.x' into image-manipulation

# Conflicts:
#	src/Facades/Glide.php
#	src/Http/Controllers/CP/Assets/ThumbnailController.php
#	src/Imaging/ImageGenerator.php

jasonvarga committed:
manipulator can be set per asset container and access through source

jasonvarga committed:
Change setParams to addParams ...

Adding focal point params is upcoming and this will make it consistent and clearer that we don't want to override the entire the entire array of params.

jasonvarga committed:
focal point params get their own method

jasonvarga committed:
use method now that the property is private

jasonvarga committed:
implement focal point method for imgix

jasonvarga committed:
initial cloudflare driver

jasonvarga committed:
fix types

jasonvarga commented:
More than likely this stale PR will be discarded in favor of something built using https://github.com/laravel/framework/pull/59276
